### ISSUE TITLE
Subscriptions using SEPA change to Manual Renewal after disabling the Legacy checkout experience

### ISSUE THREAD

**Issue by a-danae** [woocommerce/woocommerce-gateway-stripe#3133]
**Describe the bug**

Subscriptions using SEPA change to Manual Renewal after disabling the Legacy checkout experience.

**To Reproduce**

1. Check out 8.0.1. There's an error in later versions where SEPA payment methods can't be saved on the Legacy experience
1. Enable the Legacy experience
1. As a shopper, add a subscription to your cart
1. Use a SEPA src_ payment method to purchase the subscription. Like `AT611904300234573201`
1. Check out the latest version
1. As a shopper, go to My account > Subscriptions. Notice that "SEPA" is listed as the payment method
1. As a merchant, go to the Edit order page for the subscription. Notice:
    - Under payment method, it says "SEPA"
    - "Process renewal" works fine
1. Disable the legacy experience
1. As the shopper, go to My account > Subscriptions. Notice that "Manual renewal" is now displayed as the payment method for this subscription
1. As a merchant, go to the Edit order page for the subscription. Notice that:
    - Under payment method, it says "Manual renewal"
    - "Process renewal" fails

**Expected behavior**

When disabling the Legacy experience, the payment method associated with the subscription must continue to be available. Renewals must not fail.

**Additional context**

We have two different payment gateway IDs for SEPA depending on whether we use the Legacy experience:
- `stripe_sepa`, when the Legacy experience is enabled (Sources API)
- `stripe_sepa_debit` when the Legacy experience is disabled (Payment Methods API)

When purchasing a subscription using the Sources API, we set `stripe_sepa` as its associated payment method gateway.
 
<img width="400" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/04520b0a-8428-47d0-947f-7dbc5c80b3b1">

Because we use a different payment gateway ID when switching to the Payment Methods API, WooCommerce detects the `stripe_sepa` payment gateway as no longer available. This causes the Subscription to change to Manual renewal, and automatic renewals to fail.

Whether a subscription uses manual renewal is determined by the following path:

>WC_Subscription::is_manual()
>WC_Subscriptions_Payment_Gateways::has_available_payment_method()
>wc_get_payment_gateway_by_order() 
>WC_Order::get_payment_method()
>WC_Data::get_prop()

**Reports**
- 8035522-zd-a8c
- 8056842-zd-a8c
- https://wordpress.org/support/topic/no-sepa-payment-with-new-checkout-experience/
- 8104366-zd-a8c / https://wordpress.org/support/topic/upe-sepa-not-working
- 8130440-zd-a8c
- 8135622-zd-a8c

**Comment by a-danae** [comment]
Temporary workaround for most cases:

```
add_filter( 'woocommerce_subscription_get_payment_method', 'use_sepa_updated_checkout_gateway_id_for_subscriptions' );

/**
 * Filters the return value of 'WC_Subscription::get_payment_method()'.
 *
 * Returns 'stripe_sepa_debit' if UPE is enabled and the gateway is 'stripe_sepa'.
 *
 * @param string $value The payment method ID.
 * @return string
 */ 
function use_sepa_updated_checkout_gateway_id_for_subscriptions( $value ) {
	if ( 
		'stripe_sepa' === $value && 
		class_exists( 'WC_Stripe_Feature_Flags' ) && 
		WC_Stripe_Feature_Flags::is_upe_checkout_enabled() 
	) {
		return 'stripe_sepa_debit';
	}

	return $value;
}
```

**Comment by sukafia** [comment]
8162849-zen

**Comment by dsmithweb-a8c** [comment]
In 8162849-zen, user fixed issue by rolling back to Stripe 8.0.1

**Comment by agenceKanvas** [comment]
> Temporary workaround for most cases:
> 
> ```
> add_filter( 'woocommerce_subscription_get_payment_method', 'use_sepa_updated_checkout_gateway_id_for_subscriptions' );
> 
> /**
>  * Filters the return value of 'WC_Subscription::get_payment_method()'.
>  *
>  * Returns 'stripe_sepa_debit' if UPE is enabled and the gateway is 'stripe_sepa'.
>  *
>  * @param string $value The payment method ID.
>  * @return string
>  */ 
> function use_sepa_updated_checkout_gateway_id_for_subscriptions( $value ) {
> 	if ( 
> 		'stripe_sepa' === $value && 
> 		class_exists( 'WC_Stripe_Feature_Flags' ) && 
> 		WC_Stripe_Feature_Flags::is_upe_checkout_enabled() 
> 	) {
> 		return 'stripe_sepa_debit';
> 	}
> 
> 	return $value;
> }
> ```
As I have the bug in a production website, how could I help to test all that ? 
I've rolled back to 8.0.1 but the renewal order created since the update are of course still in manual, event with the quick fix. 
Fully dedicated to help you on that !

**Comment by agenceKanvas** [comment]
@a-danae If I can help with testing or anything, tell me. 
And for your workaround to work, should I keep the plugin reverted to 8.0.1 ? Or do I have to keep the last version ?

**Comment by julienvinzent** [comment]
Hi @a-danae or anyone, could we have some update or alternative workaround on this topic ?

It is 40 % of our revenue which are affected and I guess we are not alone, seeing the topics in support.
We are talking about more than 20 % of transactions in major european countries such as Germany, France, Spain or Italy...

As a WooCommerce x Stripe user since 6 years, I am more than surprised by the apparent low priority of this topic. I thinks this should at least trigger a warning in the documentation regarding this update

**Comment by a-danae** [comment]
> could we have some update or alternative workaround on this topic ?

Hey @julienvinzent! Sure, I understand how problematic this issue could be. Fully on this now.

As for workarounds, we have these two options:
a. Enabling the [Legacy checkout experience](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows-%E2%80%90-Instructions#enable-the-legacy-checkout-experience) should resume the automatic renewals while we fix this problem.
b. Or, when keeping the Legacy experience disabled, the snippet I mentioned [above](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3133#issuecomment-2107643526) should work.

There's a draft PR for solving this issue; we're currently working on it and aiming to have it out as soon as possible.

**Comment by agenceKanvas** [comment]
Thanks for your reply @a-danae ! 
For your snippet to work, should I have the last version of the plugin installed of the 8.0.1 ?
I've tried to revert to the legacy version but the form does not work (a lot of JS errors, telling that the dom is not right). 
Thanks !

**Comment by a-danae** [comment]
> And for your workaround to work, should I keep the plugin reverted to 8.0.1 ? Or do I have to keep the last version ?

Hi @agenceKanvas! It should work on all the versions after 8.0.0. I used the latest version on my tests. I'd suggest waiting for  8.3.1 to upgrade to the latest version, which should be released soon.

> As I have the bug in a production website, how could I help to test all that ?

Double checking whether I got you right. Would you like to test the snippet with the filter mentioned above, or would it be something else?

And thanks for offering a hand there!

**Comment by agenceKanvas** [comment]
Both :) 
We need a temp fix to avoid the manual renewal of course, and then I will also be able to test the 8.3.1 on our staging site if you need. 
I've tried the snippet already but it doesn't seems to change the message in the back office (in the subscription page), that tells what type of renewal is setup.

**Comment by a-danae** [comment]
> I've tried the snippet already but it doesn't seems to change the message in the back office (in the subscription page), that tells what type of renewal is setup.

@agenceKanvas Interesting. Okay, let's try a different workaround instead of diving there.

Would you please try disabling the updated checkout experience? 

1. Go to the Stripe extension settings tab, under `siteurl/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`
2. Scroll down to the "Advanced settings" section
3. If you're using 8.0.1, the checkbox should be labeled "Enable the updated checkout experience", and we'd want to uncheck it. 
   If you're on 8.1.0 or newer, it [should be labeled](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows-%E2%80%90-Instructions#enable-the-legacy-checkout-experience) "Enable the legacy checkout experience" and we'd want to check it.


Please let me know if the subscriptions renewal method goes back to SEPA after doing this!

**Comment by agenceKanvas** [comment]
Hi @a-danae I can't do that. I already tried that and when I do I have a lot of JS errors due to stripe. 
I can't remember the errors exactly but they were about the dom not being right (a problem with the exp date field i think but I don't remember).

**Comment by xoogu** [comment]
If we update the postmeta table to add in
_stripe_upe_payment_type 	sepa_debit
for all stripe_sepa orders, and then update them all to make 
_payment_method 	stripe_sepa_debit
would this be a solution?
Or could this cause problems? I assume there is something different about the way they are handled, since otherwise why would you use a different value?

**Comment by carolframen** [comment]
8293062-zen

**Comment by agenceKanvas** [comment]
Hi ! 

@a-danae Any chance that we could try something ? 

Thanks a lot !

> > I've tried the snippet already but it doesn't seems to change the message in the back office (in the subscription page), that tells what type of renewal is setup.
> 
> @agenceKanvas Interesting. Okay, let's try a different workaround instead of diving there.
> 
> Would you please try disabling the updated checkout experience?
> 
> 1. Go to the Stripe extension settings tab, under `siteurl/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`
> 2. Scroll down to the "Advanced settings" section
> 3. If you're using 8.0.1, the checkbox should be labeled "Enable the updated checkout experience", and we'd want to uncheck it.
>    If you're on 8.1.0 or newer, it [should be labeled](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows-%E2%80%90-Instructions#enable-the-legacy-checkout-experience) "Enable the legacy checkout experience" and we'd want to check it.
> 
> Please let me know if the subscriptions renewal method goes back to SEPA after doing this!

**Comment by IliaDi** [comment]
8308178-zen

**Comment by julienvinzent** [comment]
Hi, no update here, but I see on #3139 the PR is still on test.
Would you be able to give us a schedule for this fix ?
It is still hugely impacting our business and, maybe more important, undermining our support staff spirit...
Thanks for your answer

**Comment by a-danae** [comment]
> Any chance that we could try something ?

Hey @agenceKanvas! 
Sorry to hear the workaround didn't help and that going back to the legacy experience isn't an option.  There's an open PR that would fix this, but let us check another option while that's ready, and we'll get back to you.

> Would you be able to give us a schedule for this fix ?

Hi @julienvinzent!
For sure, I can see how frustrating this can be. I want to assure you we're actively on this. The related PR is still in progress. We can't promise a date there because it depends on the testing results and potential iterations of the solution. But we'll update you over here as soon as we have one or if we have an alternative in the meantime, which we're currently exploring.

Double-checking... were you able to try any of the potential temporary workarounds from [this other comment](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3133#issuecomment-2140126766)?

**Comment by a-danae** [comment]
Hi 👋  Update over here.

The initial fix for this issue has been approved and is included in the [8.5.0-alpha](https://github.com/woocommerce/woocommerce-gateway-stripe/releases/tag/8.5.0-alpha) pre-release. It's not an official release, so please treat it as an alpha if you're giving it a try.

It takes a few hours for the subscriptions to be repaired. The process runs in scheduled actions queued for an hour in the future that could take a bit more to execute.

We're planning to include [some improvements](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139#issuecomment-2184940549) to the current implementation, but the main fix is functional.

Please let us know how it goes if you decide to give it a try!
